### PR TITLE
ESM2: TransformerEngine to HuggingFace checkpoint support

### PR DESCRIPTION
**Pull request by ohadmo**
Adding TransformerEngine to HF checkpoint conversion for ESM2

## Summary by CodeRabbit

- **New Features**
  - Two-way ESM-2 checkpoint conversion via a CLI (hf-to-te, te-to-hf) with selectable models/checkpoints and output paths; adds TE→HF round‑trip export support.

- **Documentation**
  - README updated to mark HF export as supported and include a full end-to-end developer conversion and validation workflow with examples.

- **Tests**
  - New tests covering HF↔TE roundtrips, weight parity, QKV unpacking, padding/unpadding, and config consistency.

- **Chores**
  - Docker ignores split into separate HF→TE and TE→HF export directories.

### CODE REVIEW

File: models/esm2/README.md [hunk 1]
```diff
@@ -16,6 +16,93 @@ docker run --rm -it --gpus all \
   esm2 python export.py
```
Comment (pstjohn) [inline models/esm2/README.md]: does this previous command still work?

File: models/esm2/tests/test_convert_reverse.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    for name, module in model_te.named_modules():
+        if isinstance(module, nn.Linear):
+            vanilla_layers_found.append(f"Linear layer found in {name}")
+        if isinstance(module, nn.LayerNorm):
+            vanilla_layers_found.append(f"LayerNorm layer found in {name}")
+    if vanilla_layers_found:
+        print("ERROR: Found vanilla PyTorch layers in converted TE model:")
+        for error in vanilla_layers_found:
+            print(f"WARNING: {error}")
+        assert not vanilla_layers_found, f"Found {len(vanilla_layers_found)} vanilla layers in converted model"
```
Comment (pstjohn) [inline models/esm2/tests/test_convert_reverse.py]: this is a duplicate test, right? we have `test_esm_model_has_all_te_layers` in `test_modeling_esm_te`? do we need this in a test that exports from the TE model to HF?

File: models/esm2/tests/test_convert_reverse.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    assert model_hf_converted.config.num_attention_heads == model_hf.config.num_attention_heads
+    assert model_hf_converted.config.intermediate_size == model_hf.config.intermediate_size
+    assert model_hf_converted.config.vocab_size == model_hf.config.vocab_size
+
+    # assert not hasattr(model_hf_converted.config, 'qkv_weight_interleaved')
+    # assert not hasattr(model_hf_converted.config, 'encoder_activation')
+    # assert not hasattr(model_hf_converted.config, 'attn_input_format')
+    # assert not hasattr(model_hf_converted.config, 'fuse_qkv_params')
+    # assert not hasattr(model_hf_converted.config, 'micro_batch_size')
+    # assert not hasattr(model_hf_converted.config, 'max_seq_length')
```
Comment (pstjohn) [inline models/esm2/tests/test_convert_reverse.py]: any reason these are commented out?

**Review by pstjohn — APPROVED** [review]
could you make sure you rebase and run CI?

File: models/esm2/README.md (hunk 1 above)
Comment (ohadmo) [reply models/esm2/README.md]: Yes it does

File: models/esm2/tests/test_convert_reverse.py (hunk 3 above)
Comment (ohadmo) [reply models/esm2/tests/test_convert_reverse.py]: Thanks, forgot to uncomment that.  I rewrote this section

File: models/esm2/tests/test_convert_reverse.py (hunk 2 above)
Comment (ohadmo) [reply models/esm2/tests/test_convert_reverse.py]: Yes, this test is redundant, I removed it

File: models/esm2/README.md [hunk 4]
```diff
@@ -16,6 +16,93 @@ docker run --rm -it --gpus all \
...
+
+### Step 1: Load Original Hugging Face Model
+
+First, load the original ESM2 model from Hugging Face:
+
+```python
+from transformers import AutoModelForMaskedLM
+
+model_hf_original = AutoModelForMaskedLM.from_pretrained("facebook/esm2_t6_8M_UR50D")
+```
```
[the remark this answers is not in the document]
Comment (ohadmo) [reply models/esm2/README.md]: fixed

File: models/esm2/export_te_checkpoint_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Convert ESM2 models from Transformer Engine format back to HuggingFace Facebook ESM-2 format hosted on Hugging Face"
+    )
+    parser.add_argument(
+        "--model",
+        type=str,
+        help="Path to the TE checkpoint.",
+    )
```
[the remark this answers is not in the document]
Comment (ohadmo) [reply models/esm2/export_te_checkpoint_to_hf.py]: fixed

File: models/esm2/src/esm/convert.py [hunk 6]
```diff
@@ -72,6 +76,67 @@ def convert_esm_hf_to_te(model_hf: nn.Module, **config_kwargs) -> nn.Module:
...
+        "micro_batch_size",
+        "max_seq_length",
+        "model_type",
+    ]
+    for key in te_specific_keys:
+        hf_config_dict.pop(key, None)
+
+    hf_config = EsmConfig(**hf_config_dict, **config_kwargs)
+
+    with init_empty_weights():
```
[the remark this answers is not in the document]
Comment (ohadmo) [reply models/esm2/src/esm/convert.py:112]: fixed

File: bionemo-recipes/models/esm2/README.md [hunk 7]
```diff
@@ -77,17 +77,108 @@ Training recipes are available in the `bionemo-recipes/recipes/` directory:
...
 ```bash
-mkdir -p checkpoint_export
+mkdir -p hf_to_te_checkpoint_export
 docker build -t esm2 .
 docker run --rm -it --gpus all \
-  -v $PWD/checkpoint_export/:/workspace/bionemo/checkpoint_export \
+  -v $PWD/hf_to_te_checkpoint_export/:/workspace/bionemo/hf_to_te_checkpoint_export \
   -v $HOME/.cache/huggingface/:/root/.cache/huggingface \
-  esm2 python export.py
+  esm2 python export.py hf-to-te
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/README.md:85]: i don't think we need to make folders for the te-to-hf path, we just need to show how to take an existing TE model and convert it to HF (e.g., through a python code demo). Folks likely wont convert our TE models back to HF

File: bionemo-recipes/models/esm2/README.md [hunk 8]
```diff
@@ -77,17 +77,108 @@ Training recipes are available in the `bionemo-recipes/recipes/` directory:
...
+```bash
+MODEL_TAG=esm2_t6_8M_UR50D # specify which model to convert
+mkdir -p te_to_hf_checkpoint_export
+docker build -t esm2 .
+docker run --rm -it --gpus all \
+  -v $PWD/te_to_hf_checkpoint_export/:/workspace/bionemo/te_to_hf_checkpoint_export \
+  -v $PWD/hf_to_te_checkpoint_export/$MODEL_TAG:/workspace/bionemo/hf_to_te_checkpoint_export/$MODEL_TAG \
+  -v $HOME/.cache/huggingface/:/root/.cache/huggingface \
+  esm2 python export.py te-to-hf --checkpoint-path /workspace/bionemo/hf_to_te_checkpoint_export/$MODEL_TAG
+```
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/README.md:99]: i think this could be just something like
```python
from esm.convert import convert_esm_te_to_hf

te_model = AutoModel.from_pretrained("/path/to/te/checkpoint", trust_remote_code=True)
hf_model = convert_esm_te_to_hf(te_model)
hf_model.save_pretrained("/path/to/exported/model")
```

File: bionemo-recipes/models/esm2/README.md [hunk 9]
```diff
@@ -77,17 +77,108 @@ Training recipes are available in the `bionemo-recipes/recipes/` directory:
...
+### Step 2: Export to Transformer Engine Format
+
+Convert the Hugging Face model to Transformer Engine format using the high-level export API:
+
+```python
+from pathlib import Path
+from esm.export import export_hf_checkpoint
+
+te_checkpoint_path = Path("te_checkpoint")
+export_hf_checkpoint("esm2_t6_8M_UR50D", te_checkpoint_path)
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/README.md:126]: this isn't as obvious a function name as `convert_esm_te_to_hf`, let's just keep the existing API and add the reverse path?
